File: src/api/opencode.ts
import { invokeWithResponseSchema, listenWithResponseSchema } from "@/api/ipc";
import {
	type AgentRuntimeConfig,
	AgentLoginStatusSchema,
	AgentRunResultSchema,
	AgentRuntimeConfigSchema,
	AgentRuntimeStatusSchema,
} from "@/types/agent";

/** Checks OpenCode credentials and resolved runtime configuration through official CLI commands. */
const checkOpenCodeLogin = () =>
	invokeWithResponseSchema("check_opencode_login", AgentLoginStatusSchema);

/**
 * Returns the complete OpenCode status needed for the first render.
 *
 * @example
 * checkOpenCodeInitStatus();
 */
const checkOpenCodeInitStatus = () =>
	invokeWithResponseSchema(
		"check_opencode_init_status",
		AgentRuntimeStatusSchema,
	);

/**
 * Reads OpenCode's resolved model configuration without checking credentials.
 *
 * @example
 * getOpenCodeRuntimeConfig();
 */
const getOpenCodeRuntimeConfig = () =>
	invokeWithResponseSchema(
		"get_opencode_runtime_config",
		AgentRuntimeConfigSchema,
	);

/**
 * Subscribes to native changes across OpenCode's file-backed configuration layers.
 *
 * @example
 * onOpenCodeConfigChanged(refreshOpenCodeStatus);
 */
const onOpenCodeConfigChanged = (
	listener: (config: AgentRuntimeConfig) => void,
) =>
	listenWithResponseSchema(
		"opencode-config-changed",
		AgentRuntimeConfigSchema,
		listener,
	);

/**
 * Sends one task through OpenCode's documented non-interactive JSON event mode.
 *
 * @example
 * runOpenCodeTask("解释这个仓库");
 */
const runOpenCodeTask = (query: string) =>
	invokeWithResponseSchema("run_opencode_task", AgentRunResultSchema, {
		request: { query },
	});

export {
	checkOpenCodeInitStatus,
	checkOpenCodeLogin,
	getOpenCodeRuntimeConfig,
	onOpenCodeConfigChanged,
	runOpenCodeTask,
};

File: src-tauri/src/commands/opencode.rs
use crate::adapters::opencode::SystemOpenCodeAdapter;
use crate::dto::agent::{
    AgentInitStatusResponse, AgentLoginStatusResponse, AgentRunResponse, AgentRuntimeConfigResponse,
};
use crate::error::{AppError, IpcError};
use crate::services::agent::{
    check_agent_init_status, check_agent_login, load_agent_runtime_config, run_agent_task,
};
use serde::Deserialize;

/// User input accepted by the OpenCode task command.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunOpenCodeTaskRequest {
    /// Natural-language task sent to OpenCode's documented non-interactive runner.
    query: String,
}

/// Checks whether a locally installed OpenCode CLI has any usable provider credentials.
#[tauri::command]
pub async fn check_opencode_login() -> Result<AgentLoginStatusResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| check_agent_login(&SystemOpenCodeAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Returns the complete first-load OpenCode status from independent probes.
#[tauri::command]
pub async fn check_opencode_init_status() -> Result<AgentInitStatusResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| check_agent_init_status(&SystemOpenCodeAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Reads OpenCode's effective configuration without querying provider credentials.
#[tauri::command]
pub async fn get_opencode_runtime_config() -> Result<AgentRuntimeConfigResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(|| load_agent_runtime_config(&SystemOpenCodeAdapter))
        .await
        .map_err(|_| IpcError::from(AppError::WorkerFailed))?
        .map(Into::into)
        .map_err(Into::into)
}

/// Sends a bounded query through `opencode run --format json` and returns normalized metrics.
#[tauri::command]
pub async fn run_opencode_task(
    request: RunOpenCodeTaskRequest,
) -> Result<AgentRunResponse, IpcError> {
    tauri::async_runtime::spawn_blocking(move || {
        run_agent_task(&SystemOpenCodeAdapter, &request.query)
    })
    .await
    .map_err(|_| IpcError::from(AppError::WorkerFailed))?
    .map(Into::into)
    .map_err(Into::into)
}
